[Loh Jun Heng Donovan] iP

CaptainKeqing opened the pull request:
## Lindi (Log It N Do It)
Lindi: The all in one app for managing your upcoming Tasks

> "I know that I know nothing" - Socrates [source](https://en.wikipedia.org/wiki/I_know_that_I_know_nothing)

## What it can do:

- Text-based chatbot
- Task tracker
- Organiser

## What it can't do:

- Be your friend
- Do your tasks for you
- Fly a plane (~~coming soon~~)

All you need to do is,

1. Download it from [here](https://github.com/CaptainKeqing/ip/releases/tag/A-Jar)
2. Move the jar file to an empty folder of your choice
3. Run with `java -jar lindi.jar`
4. Throw away your whiteboards and notepads, Lindi is here to stay :smile: 

Best part? ITS COMPLETELY **FREE**

## Features

- [x] Track todos, deadlines and events
- [x] Mark a task as done/undone
- [x] Search for task via keywords
- [ ] Reminders sent to your mailbox (coming soon)

## For advanced users:
Change the directory path and data file name to your liking in the `main` method of Lindi.java.
```java
public static void main(String[] args) {
        new Lindi("./.data", "LindiData.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

CaptainKeqing committed:
Renamed Duke.java to Lindi.java, implemented simple greet and exit

CaptainKeqing committed:
Refactored good bye message to a function call goodByeAndExit(), which also terminates the program. Removed junk comments

CaptainKeqing committed:
Refactored greeting messages into a function call greeting().

CaptainKeqing committed:
Added chatLoop that echoes what the user types, and exists when user types 'bye'

CaptainKeqing committed:
Created function parseInputAndExecute and condensed all such parsing userInput functionality into that. Added add, list capabilities for tasks

CaptainKeqing committed:
Improved UI by improving feedback when task added

CaptainKeqing committed:
Created Task.java class and modified Lindi.java to handle the changed representation of tasks as Task not String

CaptainKeqing committed:
Added function to mark tasks as done and undone

CaptainKeqing committed:
Created classes Deadline, Event and ToDo

CaptainKeqing committed:
Removed complicated logic check for mark and unmark. Will handle them in future with error handling

CaptainKeqing committed:
Made Task abstract and factory method for the 3 subclasses

CaptainKeqing committed:
Added supported for tracking ToDos, Deadlines and Events

CaptainKeqing committed:
Improved text UI response

CaptainKeqing committed:
Tested using I/O redirection for automated testing. Changed method to receive input to use the Scanner object, as it can take in file input as well

CaptainKeqing committed:
Create Exception classes related to creation of tasks

CaptainKeqing committed:
Added error handling related to creation of tasks, as well as parsing user input

CaptainKeqing committed:
added more text feedback (when list on empty taskList), and updated automated text UI texting to reflect more cases

CaptainKeqing committed:
Changed implementation of taskList to ArrayList<Task> instead of static Task[] array

CaptainKeqing committed:
Added support for deleting tasks from the list

CaptainKeqing committed:
Fixed long standing bug for mark, unmark and delete crashing when not receiving proper arguments. Now handles the error.

CaptainKeqing committed:
Compiled functionality of mark, unmark and delete into singular function executeModifyTask. Removed redundant functions

CaptainKeqing committed:
Create primitive saveToFile() function

saveToFile() simply saves to a file in ./.data/LindiData.txt, creating
the /.data directory and txt file if not exists yet.
Currently writes a bogus message.

CaptainKeqing committed:
Add /.data/ directory to gitignore

CaptainKeqing committed:
Add parsedFormatToSave() method to Task and its subclasses

CaptainKeqing committed:
Update saveToFile() to actually save taskList information

CaptainKeqing committed:
Add createFromParsedString() method to Task

This takes in as input the parsedString that we save in the data file
and calls the Task.create() constructor. Basically we are enabling
our Task class to be loaded from the save file

CaptainKeqing committed:
Add save and load feature to Lindi

Lindi does not permanently store changes to the taskList, as it
is stored in a temporary array.

A new session would effectively wipe the users taskList. This is
not conducive for a user who wants to use the app for real life
task tracking.

Let's
* implement loading our current taskList when we start the session
* saving our taskList automatically whenever we make a change

We save it as a text in a txt file in a hidden folder in our current
directory '/.data/'. We save it as text for now because of ease of
implementation as well as it is human readable, which makes it easier
to debug.

CaptainKeqing committed:
Merge branch 'branch-Level-7'

This branch adds the save and load feature

CaptainKeqing committed:
Change representation of dates and times from String to LocalDateTime

Date and times for Deadline and Event use String, which is a simple
representation. This makes manipulation of date and times harder.

Let's use LocalDateTime to represent the by field for Deadlines and
from and to fields for Events. This allows us to use pre-written
parse and format function.

CaptainKeqing committed:
Merge branch 'branch-Level-8'

This branch changes representation of date and time in Task subclasses
to use LocalDateTime

CaptainKeqing committed:
Create Parser, Storage, TaskList, Ui classes

CaptainKeqing committed:
Create Command class to represent executable commands

Adds CreateTodoCommand, CreateDeadlineCommand, CreateEventCommand,
and DeleteTaskCommand subclasses as well

CaptainKeqing committed:
Add functionality to TaskList

CaptainKeqing committed:
Add functionality to Storage

CaptainKeqing committed:
Add 'getter' function to TaskList

CaptainKeqing committed:
Use iterator for TaskList getter, Storage saveToFile

bryanyee33 commented on src/main/java/CreateDeadlineException.java:
```diff
@@ -0,0 +1,5 @@
+public class CreateDeadlineException extends CreateTaskException {
```
Should the class names be a noun? I noticed the same issue for your 4 Exception classes.

CaptainKeqing replied:
Thank you for your insightful comments @bryanyee33 . I will review your feedback and implement the changes in a future commit !
:smile:

bryanyee33 commented on src/main/java/Lindi.java:
```diff
@@ -0,0 +1,227 @@
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Scanner;
+
```
Should there be a newline here?

bryanyee33 commented on src/main/java/Lindi.java:
```diff
@@ -0,0 +1,227 @@
+                    }
+                    Task task = taskList.get(listIndex - 1);
+                    switch (sTokens[0]) {
+                        case "mark":
+                            task.markDone();
+                            System.out.println("Nice! I've marked this task as done:\n\t" + task);
+                            break;
+                        case "unmark":
+                            task.unmarkDone();
+                            System.out.println("Nice! I've marked this task as not done yet:\n\t" + task);
+                            break;
+                        case "delete":
+                            taskList.remove(listIndex - 1);
+                            System.out.println("Okay. I've deleted this task:\n\t" + task +
+                                    "\nNow you have " + taskList.size() + " tasks in the list.");
+                            break;
+                        default:
```
Should the case clauses have indentations? I noticed the same issue in several other places too.

bryanyee33 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @Override
+    public String parsedFormatToSave() {
```
Should the method be named more sensibly? Maybe 'toParsedFormat'?

bryanyee33 commented on src/main/java/Lindi.java:
```diff
@@ -0,0 +1,227 @@
+
+    private static void parseInputAndExecute(String userInput) {
+        class LocalFuncs {
+            void executeList(){
```
Should the method be named more sensibly? Maybe 'displayList'?

bryanyee33 commented on src/main/java/Lindi.java:
```diff
@@ -0,0 +1,227 @@
+                }
+            }
+
+            void executeCreateTask(String s) {
```
Should the method be named more sensibly? Maybe just 'createTask'?

bryanyee33 commented on src/main/java/Lindi.java:
```diff
@@ -0,0 +1,227 @@
+                System.exit(0); // exit with code 0, terminates program
+            }
+
+            void executeModifyTask(String s) {
```
Should the method be named more sensibly? Maybe just 'modifyTask'?

bryanyee33 commented on src/main/java/Lindi.java:
```diff
@@ -0,0 +1,227 @@
+            case "list":
+                lf.executeList();
+                break;
+            case "unmark":
+            case "mark":
```
Should the fallthrough cases have comments?

bryanyee33 commented on src/main/java/Lindi.java:
```diff
@@ -0,0 +1,227 @@
+    /**
+     * Prints a greeting message.
+     */
+    private static void greeting() {
```
Should the method be named more sensibly? Maybe 'printGreeting'?

bryanyee33 commented on src/main/java/Lindi.java:
```diff
@@ -0,0 +1,227 @@
+
+            // Creates the file if it does not exist yet. No effect if it exists.
+            File dataFile = new File(dir, "LindiData.txt");
+            Boolean fileJustCreated = dataFile.createNewFile();
```
Should the Boolean be named more sensibly? Maybe 'isCreated'?

bryanyee33 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,132 @@
+        this.description = description;
+        this.isDone = false;
+    }
+    public static Task create(String fullString) throws CreateTaskException {
```
Should this be separated with a newline?

bryanyee33 reviewed (commented):
Overall, pretty decent in terms of code quality guidelines and coding standard violations, with a few nits on naming conventions. I like the many usages of comments! 👍

pjanthony2001 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @Override
+    public String parsedFormatToSave() {
```
Better naming would be toParsedFormat, similar to what the other reviewer said

pjanthony2001 commented on src/main/java/Lindi.java:
```diff
@@ -0,0 +1,227 @@
+
+public class Lindi {
+    private final static ArrayList<Task> taskList = new ArrayList<>(10);
+//    private final static Task[] taskList = new Task[100];
```
The comments are artifacts from previous versions maybe they should be removed

pjanthony2001 commented on src/main/java/Lindi.java:
```diff
@@ -0,0 +1,227 @@
+             * Creates one if it does not exist yet.
+             */
+            void saveToFile() {
+                try {
+                    // Creates the directory if it does not exist yet. No effect if it exists.
+                    String dataDir = "./.data";
+                    File dir = new File(dataDir);
+                    dir.mkdir();
+
+                    // Creates the file if it does not exist yet. No effect if it exists.
+                    File dataFile = new File(dir, "LindiData.txt");
```
Good placement of comments

pjanthony2001 commented on src/main/java/Lindi.java:
```diff
@@ -0,0 +1,227 @@
+        }
+        // if command called, will be index 0, rest of string in index 1
+        String[] userInputTokens = userInput.split(" ", 2); 
+        LocalFuncs lf = new LocalFuncs();
```
Maybe better naming would be localFunctions instead of lf

pjanthony2001 commented on src/main/java/Lindi.java:
```diff
@@ -0,0 +1,227 @@
+    private static void chatLoop() {
+        Scanner scanner = new Scanner(System.in);
+        String userInput;
+        while (true) { // This will not be an infinite loop, because goodByeAndExit() terminates the program when called
```
Comment should be indented properly, although trailing comments are allowed, for clarity it might be better to put the comment inside the while loop or just above it

pjanthony2001 commented on src/main/java/Lindi.java:
```diff
@@ -0,0 +1,227 @@
+    private static boolean isNumber(String s) {
+        try {
+            Integer.valueOf(s);
+            System.out.println("true number");
```
"true number / false number" is a bit vague, "Number formatted correctly/incorrectly" might be better

pjanthony2001 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,132 @@
+
+public abstract class Task { // Adapted from Course Website
+    protected String description;
+    protected boolean isDone;
```
good naming of boolean variable

pjanthony2001 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,132 @@
+        }
+        String command = sTokens[0];
+        String taskString = sTokens[1];
+        switch (command) {
+            case "todo":
+                return new ToDo(taskString);
+            case "deadline":
+                // Further split the remaining string based on /by
+                if (!taskString.contains("/by")) {
+                    throw new CreateDeadlineException(
+                            "You can't expect me to track a deadline if you don't give me the.. deadline right? haha\n" +
+                                    "please include '/by' followed by the deadline :)");
+                }
+                String[] descriptionNBy = taskString.split("/by");
+                if (descriptionNBy.length > 2) {
+                    throw new CreateDeadlineException("Don't procrastinate with multiple deadlines !! :)");
+                }
+                return new Deadline(descriptionNBy[0].trim(),
+                                    descriptionNBy[1].trim());
+            case "event":
+                // Further split the remaining string based on /from and /to
+                if (!taskString.contains("/from") || !taskString.contains("/to")) {
+                    throw new CreateEventException(
+                            "I'm not sure how you're going for an event without a start time or end time\n" +
+                                    "please include '/from' and '/to' so I can keep track for you :)");
+                }
+                if (taskString.indexOf("/from") > taskString.indexOf("/to")) {
+                    throw new CreateEventException("Sorry! /from must appear before /to :)");
+                }
+                String[] descriptionNFromNTo = taskString.split("/from |/to ");
+
+                if (descriptionNFromNTo.length > 3) {
+                    throw new CreateEventException("um.. I'm not sure how I can track for you an event that has\n" +
+                            "more than 1 start or end time! :)");
+                }
+                return new Event(descriptionNFromNTo[0].trim(),
+                                 descriptionNFromNTo[1].trim(),
+                                 descriptionNFromNTo[2].trim());
+        }
+        return null; // shouldn't come here, since we have done the check already before calling this factory function
```
Switch-case indentation needs to be corrected

pjanthony2001 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,132 @@
+     * @throws CreateTaskException if save file corrupted or edited such that format is not what is expected
+     */
+    public static Task createFromParsedString(String parsedString) throws CreateTaskException {
+        // Get trimmed tokens
+        // ToDos -> T | y or n | description
+        // Deadlines -> D | y or n | description | by
+        // Events -> E | y or n | description | from | to
```
This could be in your JavaDoc comment or maybe in a single multi-line comment instead, using /* ... */

pjanthony2001 reviewed (commented):
Overall, I think its quite well done, there were not many coding violations, but you could use the Checkstyle plugin as the other reviewer mentioned to weed out the remaining violations that are harder to catch.

CaptainKeqing committed:
Refactor Storage class and create StorageLoadException

* Storage class: now creates the data file only when saving
(if it does not exist),

* Storage: loadFromFile() throws StorageLoadException
on corrupt file or file not found

CaptainKeqing committed:
Add functionality to Ui.java

CaptainKeqing committed:
Add mark, unmark, delete, size functionality to TaskList

CaptainKeqing committed:
Implement some subclasses of Command

CaptainKeqing committed:
Remove saveToFile() throws IOException

Instead, throws RuntimeException since user is not expected
to recover from a IOException

CaptainKeqing committed:
TaskList class: Add missing return line to unmark()

CaptainKeqing committed:
Add ExitCommand and ListCommand

CaptainKeqing committed:
Implement some functionality in parser

Now supports list, mark, unmark, delete, bye
Fixes some printing in Ui

CaptainKeqing committed:
Move logic to the respective OOP classes

CaptainKeqing committed:
Event and Deadline: Fix error msg showing wrong format

CaptainKeqing committed:
Unite task creation commands under CreateTaskCommand

CaptainKeqing committed:
Add support for task creation, add null check

Null checks added for prepareMark, prepareUnmark and prepareDelete

CaptainKeqing committed:
Improve javadocs coverage

CaptainKeqing committed:
Add javadocs, improve user feedback for load error

CaptainKeqing committed:
Merge branch 'branch-A-MoreOOP'

CaptainKeqing committed:
Organise files into appropriate packages

CaptainKeqing committed:
Merge branch 'branch-A-Packages'

This branch organises the java files into appropriate packages.

CaptainKeqing committed:
Improve javadocs for commands package

CaptainKeqing committed:
Follow coding standard for parser, storage and task packages

CaptainKeqing committed:
Create FindCommand, add description in Task to support keyword search

CaptainKeqing committed:
Implement find command functionality

CaptainKeqing committed:
Improve Javadocs for parser and storage packages

CaptainKeqing committed:
Make datetime related format variables in Task static

CaptainKeqing committed:
Improve Javadocs for task package

CaptainKeqing committed:
Merge branch 'branch-A-JavaDoc'

This branch improves the existing javadoc coverage in the spirit
or better documentation

CaptainKeqing committed:
Merge branch 'branch-A-CodingStandard'

This branch modifies the code to better follow the
coding standard given to us.

CaptainKeqing committed:
Merge branch 'branch-Level-9'

This branch implements find feature.

CaptainKeqing committed:
Merge branch 'add-gradle-support'

CaptainKeqing committed:
Add gradle support and linked to application

CaptainKeqing committed:
Follow more coding standard conventions

CaptainKeqing committed:
Enable assertions for gradle run

CaptainKeqing committed:
Add Junit tests for Parser

CaptainKeqing committed:
Add Junit test for Storage saveToFile empty tasks

CaptainKeqing committed:
Fix checkstyle for test files

CaptainKeqing committed:
Merge branch 'branch-A-Checkstyle'

CaptainKeqing committed:
Implement cosmestic and code quality feedback from PR review

CaptainKeqing committed:
Add JavaFx support to gradle, start JavaFx tutorial

CaptainKeqing committed:
Implement JavaFx GUI to replace text based ui

This commit provides the base to work off to improve
the user interface by using JavaFx for GUI.

This includes some refactors to the code. Notably, the
entry into the application will now be Launcher.main().

CaptainKeqing committed:
Merge branch 'branch-Level-10'

CaptainKeqing committed:
Make application non resizable

CaptainKeqing committed:
Change output jar name

CaptainKeqing committed:
Merge pull request #1 from CaptainKeqing/branch-Level-10

Branch level 10

CaptainKeqing committed:
Add assertion to create() method

The create() method has assumptions about the length
of the split input string which are true after a series of checks.
These are however not very explicit and may break if some checks
are rewritten.

Adding asserts will make these conditions more explicit to the developer
as well as ensure they are maintained.

CaptainKeqing committed:
Add assertions to saveFromFile and loadFromFile methods

This makes necessary true conditions more explicit.

CaptainKeqing committed:
Refactor long functions to follow SLAP, improve code quality

CaptainKeqing committed:
Fix bug where error msg prints DateTime

CaptainKeqing committed:
Remove unused JavaFx import

CaptainKeqing committed:
Refactor methods in Task class, reduce method length

The create and createFromParsedString methods were too long and
involved multiple levels of abstraction.

Extracting some smaller methods out of these improves the code quality.

As a move for better code quality throughout, let's extract the
creation of the various subclasses into separate methods of their
own. Doing so will make the code more readable and less error prone.

CaptainKeqing committed:
Merge pull request #3 from CaptainKeqing/branch-A-CodeQuality

Branch to improve code quality

CaptainKeqing committed:
Merge branch 'master' into branch-A-Assertions

CaptainKeqing committed:
Merge pull request #4 from CaptainKeqing/branch-A-Assertions

Branch to add assertions

CaptainKeqing committed:
Break up CreateTaskCommand into separate commands

CaptainKeqing committed:
Break up Create factory method into seperate methods

CaptainKeqing committed:
Abstract out command input word into respective Command subclasses

The Parser checks if the user input matches certain keywords and
calls the corresponding command classes. This is hard-coded in the
parser.

Abstracting this command keywords out to the respective Command
subclasses makes sense, reducing potential human error.

Let's put the keywords in the Command subclasses, and have the
Parser access them.

CaptainKeqing committed:
Add support for mass operations for mark, unmark and delete

Currently, operations can only be executed in isolation, that is,
one index at a time.

To support operation on multiple indices, we need the addition of
a new Command subclass MassOpsCommand.

Let's
* Create the Command subclass MassOpsCommand.
* Modify the Parser to support the creation of the MassOpsCommand
to allow mass operations.

CaptainKeqing committed:
Fix test imports

CaptainKeqing committed:
Merge branch 'branch-C-MassOps'

This branch adds support for mass operations for
mark, unmark and delete

CaptainKeqing committed:
Remove unused import in DialogBox

CaptainKeqing committed:
Add javadocs for MassOpsCommand

CaptainKeqing committed:
Refactor handleUserInput() by extracting into smaller methods, follow SLAP

CaptainKeqing committed:
Add exit functionality to GUI on command 'bye'

CaptainKeqing committed:
Update user guide, add title to GUI
